File: simple-statemachine/simple-statemachine-java/src/test/java/com/example/statemachine/simple/StateMachineUnitTestHelper.java
package com.example.statemachine.simple;

import org.springframework.messaging.support.MessageBuilder;
import org.springframework.statemachine.StateMachine;
import org.springframework.statemachine.test.StateMachineTestPlan;
import org.springframework.statemachine.test.StateMachineTestPlanBuilder;
import reactor.core.publisher.Mono;

public class StateMachineUnitTestHelper<S, E> {

  private final StateMachine<S, E> stateMachine;

  public StateMachineUnitTestHelper(StateMachine<S, E> stateMachine) {
    this.stateMachine = stateMachine;
  }

  public StateMachineTestPlan<S, E> getStateMachineTestPlan(final S state) {
    return StateMachineTestPlanBuilder.<S, E>builder()
      .stateMachine(stateMachine)
      .step()
      .expectState(state)
      .and()
      .build();
  }



  public StateMachineTestPlan<S, E> getStateMachineTestPlan(final E event, final S prevState, final S expectedState) {
    return StateMachineTestPlanBuilder.<S, E>builder()
      .stateMachine(stateMachine)
      .step()
      .expectState(prevState)
      .and()
      .step()
      .sendEvent(event)
      .and()
      .step()
      .expectState(expectedState)
      .and()
      .build();
  }



  public void sendEvent(final E event) {
    final var message = Mono.just(MessageBuilder.withPayload(event)
      .build());

    stateMachine.sendEvent(message)
      .blockLast();
  }

}

File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-kotlin/src/test/kotlin/com/example/statemachine/simple/StateMachineUnitTestHelper.kt
package com.example.statemachine.simple

import org.springframework.messaging.support.MessageBuilder
import org.springframework.statemachine.StateMachine
import org.springframework.statemachine.test.StateMachineTestPlan
import org.springframework.statemachine.test.StateMachineTestPlanBuilder
import reactor.core.publisher.Mono

class StateMachineUnitTestHelper<S, E>(private val stateMachine: StateMachine<S, E>) {
  fun getStateMachineTestPlan(state: S): StateMachineTestPlan<S, E> {
    return StateMachineTestPlanBuilder.builder<S, E>()
      .stateMachine(stateMachine)
      .step()
      .expectState(state)
      .and()
      .build()
  }

  fun getStateMachineTestPlan(
    event: E,
    prevState: S,
    expectedState: S
  ): StateMachineTestPlan<S, E> {
    return StateMachineTestPlanBuilder.builder<S, E>()
      .stateMachine(stateMachine)
      .step()
      .expectState(prevState)
      .and()
      .step()
      .sendEvent(event)
      .and()
      .step()
      .expectState(expectedState)
      .and()
      .build()
  }

  fun sendEvent(documentId: Int, event: E) {
    val message = Mono.just(
      MessageBuilder.withPayload(event!!)
        .setHeader("document", documentId)
        .build()
    )


    stateMachine.sendEvent(message)
      .blockLast()
  }
}
